### DISCUSSION TITLE
Polyfills

### DISCUSSION THREAD

**Discussion by tj-kev** [flareact/flareact#107] [Q&A] [answered]
Hey,
i'm trying to get polyfills to work. This is specifically to make it work within IE11, this is a requirement for our project . I get the following error in the IE11 console:
```
SCRIPT5009: 'Promise' is undefined
```

I have tried the normal polyfill way for react:
```
import 'react-app-polyfill/ie11';
import 'react-app-polyfill/stable';
```

This seems to break things further though? Nothing loads but looking in the CF worker quick edit console to manually execute the worker I see:

```
Failed to execute 'respondWith' on 'FetchEvent': parameter 1 is not of type 'Promise'.
```

Any suggestions for this? I've noticed the Flareact documentation site also suffers from this issue. Which stops the search field from working.

**Comment by tj-kev** [comment]
Any body have an suggestions for this? Its a show stopper for us being able to use this framework.

**Comment by jplhomer** [comment]
Hi @tj-kev! Sounds tricky, but you might investigate loading a polyfill as a `public/polyfills.js` script, as described here: https://webpack.js.org/guides/shimming/#loading-polyfills

Another thing to play with is to hook into the Webpack bundler and "inject" polyfills somehow - maybe via global shimming: https://flareact.com/docs/custom-webpack-config

Note that you can choose to inject something based on whether `isWorker` is truthy, so it only gets included in the client bundle.

Finally, something to watch out for: most polyfills and examples will attempt to key off of `typeof window !== undefined` or `window.Promise` - but both of those are present in the Workers runtime 🙈 So you might have to rely on something else!

**Reply by jplhomer** [reply]
For context: Flareact uses `fetch` in the client for getting props, `Promise` in a bunch of places, and `Set` and `Map` for the router - so you'll want to polyfill those things, too, if they're missing in IE11.

**Comment by tj-kev** [comment, marked as answer]
I have a solution which seems to be working so far.

Created a new polyfills file with all the required imports to fix IE11.
```
import 'core-js/es/promise';
import 'core-js/es/symbol'
import 'core-js/es/array';
import 'core-js/es/map';
import 'core-js/es/set';
import 'core-js/es/object/entries';
import 'core-js/es/string/ends-with';
import 'core-js/es/string/virtual/match-all';
import 'core-js/es/object/from-entries';
import 'core-js/features/url';
import 'intersection-observer';
import 'whatwg-fetch'

import 'flareact/src/client/index';
```

In that file I also import the main client index. The webpack config is overridden to set the polyfills file as the entry instead of the client index.
```
module.exports = {
    webpack: (config, { dev, isWorker, defaultLoaders, webpack }) => {
        if (!isWorker) {
            config.entry.main = process.cwd() + "/polyfills"
        }

        return config;
    },
};
```

Only added the polyfills required so far. More may be needed depending on what features are used in the future.

**Comment by iamspark1e** [comment]
In addition, if you just need a polyfill for different browser enviroment, why not try polyfill.io, it will detect `user-agent` to serve different polyfill pack, "This endpoint responds with a JavaScript file containing the polyfills which should be served to the requesting browser."

**Reply by tj-kev** [reply]
Tried that but could never get it to work properly. I think because react was injecting it after itself had loaded. Couldn't find away around it.

Also for other reasons I'm trying to keep the third party scripts to a minimum. But if that is not a consideration and you can get it to work then its definitely an option.
